cmd/kubeadm/app/util/apiclient/wait.go

```diff
@@ -186,11 +186,17 @@ func (w *KubeWaiter) WaitForStaticPodControlPlaneHashes(nodeName string) (map[st
 
 	componentHash := ""
 	var err error
+	var lastErr = errors.New("")
 	mirrorPodHashes := map[string]string{}
 	for _, component := range kubeadmconstants.ControlPlaneComponents {
+		staticPodName := fmt.Sprintf("%s-%s", component, nodeName)
 		err = wait.PollImmediate(kubeadmconstants.APICallRetryInterval, w.timeout, func() (bool, error) {
-			componentHash, err = getStaticPodSingleHash(w.client, nodeName, component)
+			componentHash, err = getStaticPodSingleHash(w.client, staticPodName)
 			if err != nil {
+				if err.Error() != lastErr.Error() {
+					fmt.Printf("[apiclient] Get static pod %s hash failed with err: %v\n", staticPodName, err)
+					lastErr = err
+				}
```

Comment:
try:
```
	var err, lastErr error
	mirrorPodHashes := map[string]string{}
	for _, component := range kubeadmconstants.ControlPlaneComponents {
		err = wait.PollImmediate(kubeadmconstants.APICallRetryInterval, w.timeout, func() (bool, error) {
			componentHash, err = getStaticPodSingleHash(w.client, nodeName, component)
			if err != nil {
				lastErr = err
				return false, nil
			}
			return true, nil
		})
		if err != nil {
			if lastErr != nil {
				err = lastErr
			}
			return nil, errors.Wrapf(err, "failed to obtain static Pod hash for component %s on Node %s, component, node)
		}
```

similar for WaitForStaticPodSingleHash

Reply:
Thanks! seems good, but a little doubt that this may lead user not be aware err so fast, it will return err message until timeout.

Reply:
kubeadm has a few places where it prints errors during polls iterations with higher verbosity (v >= 4), "etcd member join", "node join" come to mind, but the other places mostly print the last error after the poll.

this particular timeout is long (5 minutes)...:
https://github.com/kubernetes/kubernetes/blob/cde45fb161c5a4bfa7cfe45dfd814f6cc95433f7/cmd/kubeadm/app/phases/upgrade/staticpods.go#L49

i'd remove this message because it's redundant:
https://github.com/kubernetes/kubernetes/blob/1fa888529e398424ab0823508e01826f0a70dccb/cmd/kubeadm/app/cmd/upgrade/apply.go#L154

and modify this message to include the timeout to let the users know how long it is:
https://github.com/kubernetes/kubernetes/blob/1fa888529e398424ab0823508e01826f0a70dccb/cmd/kubeadm/app/cmd/upgrade/apply.go#L225

```
fmt.Printf("[upgrade/apply] Upgrading your Static Pod-hosted control plane to version %q (timeout: %v)...\n",
    internalcfg.KubernetesVersion, waiter.timeout)
```

the retry interval is 500ms so it will be very spammy if we print each error for 5 minutes.
https://github.com/kubernetes/kubernetes/blob/f11dbaea2f2b62f43144b8f8c8f645fb0a6eeb4f/cmd/kubeadm/app/constants/constants.go#L210

we don't do it in other places:
https://github.com/kubernetes/kubernetes/blob/9804a83d8fd9eec4c20e75e73e11ffd4b370c6f0/cmd/kubeadm/app/util/apiclient/idempotency.go

Reply:
> try:
> 
> ```
> 	var err, lastErr error
> 	mirrorPodHashes := map[string]string{}
> 	for _, component := range kubeadmconstants.ControlPlaneComponents {
> 		err = wait.PollImmediate(kubeadmconstants.APICallRetryInterval, w.timeout, func() (bool, error) {
> 			componentHash, err = getStaticPodSingleHash(w.client, nodeName, component)
> 			if err != nil {
> 				lastErr = err
> 				return false, nil
> 			}
> 			return true, nil
> 		})
> 		if err != nil {
> 			if lastErr != nil {
> 				err = lastErr
> 			}
> 			return nil, errors.Wrapf(err, "failed to obtain static Pod hash for component %s on Node %s, component, node)
> 		}
> ```
> 
> similar for WaitForStaticPodSingleHash

Thanks, and maybe there is no need to check `lastErr != nil` because when `err!=nil` it must be timeout err here and `lastErr` is not nil either, so we can return `lastErr` directly?

Reply:
i think you are right and we don't need it in this case.
if `err != nil`, `lastErr` will always be `!= nil` as well.

Reply:
note, some of the parts mentioned here:
https://github.com/kubernetes/kubernetes/pull/108315#discussion_r814856146
are still missing in the diff.

Reply:
Thanks, it's done. 
And also `WaitForStaticPodHashChange` is a little special, we should distinguish two errors because it maybe apiclient getStaticPodSingleHash error or pod hash didn't change err.